### DISCUSSION TITLE
Shortcut for 'Run on Device'

### DISCUSSION THREAD

**Discussion by birdistheword96** [pycom/pymakr-vsc#233] [Ideas]
![image](https://user-images.githubusercontent.com/53791523/168775675-c9c377f3-7bfa-4fe4-ac33-1a475712019e.png)

There is already a shortcut for 'Upload to Device', but often if you're making quick changes to a screen gui or a scripts behaviour in the main.py file, you just want to run the script and check the changes. Having a shortuct for that would improve the workflow massively, and allow developers to quickly test there ideas without having to right-click, find pymaker list item, then click on 'run on device', then select the device, then hit enter every time.

Bonus would also be to auto-select the device and immediately run it if you only have one connected, that would save even more time!

**Comment by Josverl** [comment]
from the manifest there should already be a default shortcut defined :
```
        "command": "pymakr.runEditor",
        "key": "ctrl+shift+r",
        "mac": "cmd+shift+r"
```
naturally it is possible to add additional/different keys to the command as you prefer 
File > preferences > Keyboard Shortcuts > Search for the command `pymakr.run`, and add your own hotkeys 

I fully agree that it should default to running on the last active (or only) device 
also that terminal should be made visible if it is not, so that all output is can be seen.

If there is a (small) selection , then that command can be run directly in the terminal as that then allow one to simply inspect variables by selecting and running them.

**Comment by jakobrosenberg** [comment]
> also that terminal should be made visible if it is not, so that all output is can be seen.

I'm for this as well. We can also stream all device data to the output tab, so that each device gets its own entry.

> it should default to running on the last active (or only) device

As for the behavior of running a script, I think all use cases have to be considered before we make any changes to existing functionality.

Say I have a workspace with a couple of pymakr projects. The projects are related, but require different devices. If I run a file from project A I would expect it to run on one of the compatible devices from project A.

Currently, running a script on a device can be done with `ctrl + shift + r` -> `enter`. The only way to shorten it is to remove the `enter` part.

There are a few ways to approach this:
**Option A)** Remove `enter` and choose for the user. This would prevent users from having a say in whether they want to target all or just one device. For a large batch job you might want to select 10s of devices and for a quick test, just one.
**Option B)** Create a separate `quick-run` option that targets a single device.
**Option C)** Move the checkboxes to the project devices view. This last option has a few benefits. Devices in the project view can be marked as selected and this will affect whether they're affected by actions such as `upload` and `run`. This would eliminate the need for a select-devices-dialog.

**Reply by Josverl** [reply]
I was considering the below approach :
run a selection of code should  should be as easy as Python `Shift-Enter` (or any key-chord of choice) 
as PyMakr can have multiple project and devices active there should be some logic to what is run where.

I think that could be that quick-run option
> Create a separate quick-run option that targets a single device.

Currently every decision-branch requires a prompt, with sensible defaults, but not always very clear 
the goal would be to skip prompts, if possible:
- for : "run a file or selection"
  - is file in project ?
      - get devices for project 
      - if more than one device configured :   Ask which device to use 
         otherwise use the single device 
  - if file is not in project 
   -   if one device is connected , use that device 
   - in multiple devices connected , prompt for device to use 
 
after a device has been defaulted or selected : 
  -  if multi-line selection : run though raw paste 
   - else : run though simple paste to show output of simple statements without needing print

**Reply by Josverl** [reply]
> For a large batch job you might want to select 10s of devices and for a quick test, just one.


That is an interesting scenario that I had not even considered. 👍🏼👍🏼
I would definitly describe this in the documentation

**Comment by sebi5361** [comment]
@birdistheword96 said:
> Bonus would also be to auto-select the device and immediately run it if you only have one connected, that would save even more time!

@Josverl said:
> If there is a (small) selection , then that command can be run directly in the terminal as that then allow one to simply inspect variables by selecting and running them.

The **Autoselect** feature would be neat indeed, as, I believe, the default scenario will be people working on one single board at a time only.
PS: Nice work! I like where this new version of Pymakr is going!!

**Comment by jakobrosenberg** [comment]
New Preview version out. `2.18.0`

Running scripts with `ctrl + shift + r`
![run-script](https://user-images.githubusercontent.com/4153004/173809056-4f27d7d1-5705-4325-804c-a385bc515471.gif)
